src/goto-instrument/contracts/contracts.cpp

```diff
@@ -246,6 +279,18 @@ void code_contractst::check_apply_loop_contracts(
     goto_function.body.destructive_insert(std::next(loop_head), havoc_code);
   }
 
+  // Reassign 'loop_old' history variables before the loop body
+  std::map<exprt, exprt> parameter2history_loop_old_2;
+  replace_history_parameter(
```

Comment:
This doesn't look right.

1. It looks like you are replacing `__CPROVER_loop_old` within invariant_copy again, which would actually not replace anything, since those have been replaced before.

2. You should generate assignments here, not replacement. Basically, for each expression in `parameter2history_loop_old` (not `parameter2history_loop_old_2`, which shouldn't be needed) you should generate an assignment statement to update the stored value with the current value of the expression.